firedrake/preconditioners/asm.py

```diff
@@ -0,0 +1,98 @@
+import abc
+
+from pyop2.mpi import COMM_SELF
+from firedrake.preconditioners.base import PCBase
+from firedrake.petsc import PETSc
+from firedrake.dmhooks import get_function_space
+
+__all__ = ("ASMPatchPC", "ASMLinesmoothPC")
+
+
+class ASMPatchPC(PCBase):
+    ''' PC for PETSc PCASM
+
+    should implement:
+    - :meth:`get_patches`
+    '''
+
+    _prefix = "pc_asm_"
+
+    def initialize(self, pc):
+        # Get context from pc
+        _, P = pc.getOperators()
+        dm = pc.getDM()
+        self.prefix = pc.getOptionsPrefix() + self._prefix
+
+        # Extract function space and mesh to obtain plex and indexing functions
+        V = get_function_space(dm)
+
+        # Obtain patches from user defined funtion
+        ises = self.get_patches(V)
+
+        # Create new PC object as ASM type and set index sets for patches
+        asmpc = PETSc.PC().create(comm=pc.comm)
+        asmpc.incrementTabLevel(1, parent=pc)
+        asmpc.setOptionsPrefix(self.prefix + "_sub")
+        asmpc.setOperators(*pc.getOperators())
+        asmpc.setType(asmpc.Type.ASM)
+        asmpc.setASMLocalSubdomains(len(ises), ises)
+        asmpc.setFromOptions()
+        self.asmpc = asmpc
+
+    @abc.abstractmethod
+    def get_patches(self, V):
+        ''' Get the patches used for PETSc PSASM
+        '''
+        pass
+
+    def update(self, pc):
+        pass
+
+    def apply(self, pc, x, y):
+        self.asmpc.apply(x, y)
+
+    def applyTranspose(self, pc, x, y):
+        self.asmpc.applyTranspose(x, y)
+
+
+class ASMLinesmoothPC(ASMPatchPC):
+    '''Linesmoother PC for extruded meshes implemented as an ASMPatchPC.
+
+    These patches are made by looping over base mesh entities of a
+    particular codimension, and collecting all the DOFs on the
+    "horizontal" faces.  Then looping over all facet DOFs and
+    collecting the DOFs on the "vertical" faces.
```

Comment:
I don't think I understand this statement (it's controlled by what codim_list is, no?).

I think you want to say something like, patches are built of dofs that live on the same base mesh entity (stacked in a column). To select which base entities are considered use `pc_linesmooth_codims`. For example `pc_linesmooth_codims 0,1` creates patches for each cell and each facet of the base mesh.

Reply:
But probably you can come up with a better explanation.

Reply:
Hopefully I have done.